README.md
# Victory Native

[![Maintenance Status][maintenance-image]](#maintenance-status)

### Usage

Install victory-native:
```sh
$ npm install victory-native --save
```

Install react-native-svg:
```sh
$ npm install react-native-svg --save
```

Link react-native:
```sh
$ react-native link react-native-svg
```
**`victory-native@^33.0.0` requires `react-native-svg@^9.0.0` and `react-native@~0.60.0`**

**Please see [Peer Dependencies and Version Requirements](#peer-dependencies-and-version-requirements) for requirements for previous versions of `victory-native`**


Import charts from `victory-native`. For example,

```jsx
import React, { Component } from "react";

import { VictoryBar } from "victory-native";

class App extends Component {
  render() {
    return (
      <VictoryBar />
    );
  }
}

export default App;
```

### Peer Dependencies and Version Requirements

**Note:** `victory-native` requires the following peer dependencies:
  - `react-native-svg`
  - `react`
  - `react-native`

**Note:** `react-native-svg` has strict version requirements for both `react` and `react-native`. Please match versions to those required by `react-native-svg`. See the up-to-date requirements on the [react-native-svg Readme][react-native-svg-readme].
We encourage you to use the latest version of `react-native-svg` possible for your project, as `victory-native` issues are frequently solved by `react-native-svg` bugfixes.

- `victory-native@^33.0.0` requires `react-native-svg@^9.0.0` and `react-native@~0.60.0`
* `victory-native@^30.0.0` requires `react-native-svg@6.1.x`  or `react-native-svg@^6.5.0`and above
* `victory-native@^0.16.2` requires `react-native-svg@6.1.x` or `react-native-svg@^6.5.0`
* ~~`victory-native@~0.16.0` requires `react-native-svg@6.0.0`~~ No longer supported
* ~~`victory-native@~0.15.0` requires `react-native-svg@^5.0.0`~~ No longer supported

### Local Development and Demo

If you'd like to contribute to `victory-native`, you can use the local demo app to test your changes on the iOS simulator.

```sh
# Install
$ git clone https://github.com/FormidableLabs/victory-native
$ cd victory-native
$ yarn install
# To run our typescript demo: cd demo/ts
$ cd demo/js
$ yarn install
$ yarn start

# Run the demo from a new terminal window
$ npx react-native run-ios
```

Changes to `lib` will be reflected in the demo app.

### Documentation

See the docs and examples on the website https://formidable.com/open-source/victory/docs/native

## Contributor Covenant Code of Conduct

Please review our [Code of Conduct][code] before contributing.


### Maintenance Status

**Active:** Formidable is actively working on this project, and we expect to continue for work for the foreseeable future. Bug reports, feature requests and pull requests are welcome.


[code]: https://github.com/FormidableLabs/.github/blob/master/CODE_OF_CONDUCT.md
[victory-native-demo]:https://github.com/FormidableLabs/victory-native-demo
[react-native-svg-readme]: https://github.com/react-native-community/react-native-svg#notice
[maintenance-image]: https://img.shields.io/badge/maintenance-active-green.svg



CHANGELOG.md
# VictoryNative Changelog

## 35.5.5 (2021-09-17)

- [#689](https://github.com/FormidableLabs/victory-native/pull/689) - unpin `victory` versions

## 35.5.4 (2021-08-26)

- [#676](https://github.com/FormidableLabs/victory-native/pull/676) - Pin `victory` to `35.10.x` as a bandaid until we get a full bugfix for `35.11.1`.

## 35.3.3 (2021-05-05)

- [#639](https://github.com/FormidableLabs/victory-native/pull/639) - Fixes a bug in native-specific types for container components

## 35.3.2 (2021-04-08)

- [#633](https://github.com/FormidableLabs/victory-native/pull/633) - Fixes a bug in native-specific types for `VictoryContainer`

## 35.3.1 (2020-11-10)

- [#611](https://github.com/FormidableLabs/victory-native/pull/611) - Adds a temporary fix for a bug in `react-native-svg` that is causing events on the parent Svg to not be triggered in some circumstances

## 35.3.0 (2020-11-02)

Updates all `victory-*` components and adds `accessibleLabel` attributes to primitive components that match `aria-label` props.

## 35.0.1 (2020-07-29)

Updates all `victory-*` components and adds support for new OUIA props on `VictoryContainer`

## 35.0.0 (2020-06-27)

### Label and Theme Improvements!

This release introduces new label features and makes improvements to themes. This release includes breaking style changes. Please double check your label and tooltip styles when upgrading. Polar axes and pie charts may be particularly impacted.

Please see the [main project changelog](https://github.com/FormidableLabs/victory/blob/main/CHANGELOG.md#3500-2020-06-27) for full details

## 34.3.0 (2020-05-04)

- [#565](https://github.com/FormidableLabs/victory-native/pull/565) - Adds `VictoryHistogram`. Thanks @NgoKnows!
- [#564](https://github.com/FormidableLabs/victory-native/pull/564) - Adds `Background` primitive for `VictoryChart`. Thanks @maddles and @wsparsons!


## 34.1.0 (2020-02-13)

- [#533](https://github.com/FormidableLabs/victory-native/pull/533) - Adds typescript type definitions. Thanks @mariano-formidable!
- [#538](https://github.com/FormidableLabs/victory-native/pull/538) - Adds typescript demos. Thanks @mariano-formidable!

## 34.0.0 (2019-12-23)

This PR updates `victory-*` dependencies to version `^34.0.0`. This version uses the context API introduced in `react@16.3.0`

## 33.0.1 (2019-12-01)

- [#516](https://github.com/FormidableLabs/victory-native/pull/516) - copy `role` from `VictoryLabel`. Thanks @gurglet!

## 33.0.0 (2019-08-22)

This PR updates `victory-*` dependencies to version `^33.0.0`. Please see [release notes] (https://github.com/FormidableLabs/victory/releases/tag/v33.0.0) for breaking changes

In addition, this PR adds improvents that require `victory-native@33.0.0` to use `react-native@~0.60.0` and `react-native-svg@^9.0.0`

Infrastructure, performance, and event improvements. Thanks @carlos-kelly!
- [#487](https://github.com/FormidableLabs/victory-native/pull/487) - Upgrade to RN~0.60.0
- [#490](https://github.com/FormidableLabs/victory-native/pull/490) - Fix native scroll bugs
- [#491](https://github.com/FormidableLabs/victory-native/pull/491) - Improve performance with functional primitive components and a memoized `getNativeStyles` hook
- [#496](https://github.com/FormidableLabs/victory-native/pull/496) - Ensures that scroll events are only blocked when appropriate (_e.g._ zooming)

## 32.0.2 (2019-04-01)

- [#451](https://github.com/FormidableLabs/victory-native/pull/451) Updates the demo to use `react-native@~0.59.0`. Thanks @ratson
- [#452](https://github.com/FormidableLabs/victory-native/pull/452) Fixes crashes with react-native-svg v7.0.0 - v9.2.3. Thanks @msand
- [#456](https://github.com/FormidableLabs/victory-native/pull/456) Adds a work around for a clipPath caching issue

## 32.0.1 (2019-03-18)

- [#450](https://github.com/FormidableLabs/victory-native/pull/450)updates lodash to fix a security vulnerability

## 32.0.0 (2019-03-15)

- Updates to `victory@32.0.0`. Changes described here: https://github.com/FormidableLabs/victory/blob/main/CHANGELOG.md#3200-2019-02-27

## 31.0.0 (2018-11-10)

- [victory/1177](https://github.com/FormidableLabs/victory/pull/1177) Adds support for controlling `radius`, `innerRadius`, `cornerRadius`, `padAngle`, `sliceStartAngle` and `sliceEndAngle` for each individual slice of a pie:

Details:
The `Slice` primitive used by `VictoryPie` now takes `radius`, `cornerRadius`, `innerRadius`, `padAngle`, `sliceStartAngle` and `sliceEndAngle` props. Each of these props may be given as number or a function of `datum` and `active`.  `padAngle`, `radius`, `cornerRadius` and `innerRadius` will be passed down from props on `VictoryBar`, but `sliceStartAngle` and `sliceEndAngle` must be defined directly on the `Slice` instance, These values should be given in degrees.  To make these values easier to use as functional props, `startAngle`, `endAngle`, and `padAngle` (in degrees) are added to each `datum` passed into `Slice`. (If your data already has these properties they will _not_ be overridden)

*Breaking Changes*
The `Slice` primitive will still take a `pathFunction` prop, but this prop will no longer be provided by `VictoryPie`. This will not be a breaking change for most users. This will only affect users who were  wrapping the `Slice` component and making use of the `pathFunction` prop provided by `VictoryPie` Users who were providing their own `pathFunction` prop to `Slice` should not be effected.

## 30.6.0 (2018-10-27)

- [#399](https://github.com/FormidableLabs/victory-native/pull/399) - Activate cursor `onTouchStart`

- [victory/#1158](https://github.com/FormidableLabs/victory/pull/1158) `cornerRadius` improvements for `VictoryBar`
  - Artifacts caused by `cornerRadius` values larger than the height of the bar have been corrected

demo/index.ios.js
import random from "lodash.random";
import range from "lodash.range";
import React, { Component } from "react";
import {
  AppRegistry,
  ScrollView,
  StyleSheet,
  Text
} from "react-native";
import Svg from "react-native-svg";
import {
  VictoryArea,
  VictoryAxis,
  VictoryBar,
  VictoryCandlestick,
  VictoryChart,
  VictoryGroup,
  VictoryLine,
  VictoryPie,
  VictoryScatter,
  VictoryStack,
  VictoryErrorBar,
  VictoryVoronoiTooltip,
  VictoryTooltip
} from "../lib";

const styles = StyleSheet.create({
  container: {
    alignItems: "center",
    backgroundColor: "#e1d7cd",
    justifyContent: "center",
    paddingLeft: 50,
    paddingRight: 50,
    paddingTop: 50
  },
  text: {
    fontSize: 18,
    fontFamily: "Menlo",
    fontWeight: "bold",
    marginTop: 10,
    marginBottom: 30
  }
});

const candleData = [
  {x: 1, open: 9, close: 30, high: 56, low: 7},
  {x: 2, open: 80, close: 40, high: 120, low: 10},
  {x: 3, open: 50, close: 80, high: 90, low: 20},
  {x: 4, open: 70, close: 22, high: 70, low: 5},
  {x: 5, open: 20, close: 35, high: 50, low: 10},
  {x: 6, open: 35, close: 30, high: 40, low: 3},
  {x: 7, open: 30, close: 90, high: 95, low: 30},
  {x: 8, open: 80, close: 81, high: 83, low: 75}
];

class Demo extends Component {
  constructor(props) {
    super(props);
    this.state = {
      y: this.getYFunction(),
      style: this.getStyles(),
      transitionData: this.getTransitionData(),
      randomData: this.generateRandomData()
    };
  }
  getYFunction() {
    const n = random(2, 7);
    return (data) => Math.exp(-n * data.x) * Math.sin(2 * n * Math.PI * data.x);
  }

  generateRandomData() {
    return range(1, 7).map(() => ({x: " ", y: random(1, 10)}));
  }

  getStyles() {
    const colors = [
      "red", "orange", "magenta",
      "gold", "blue", "purple"
    ];
    return {
      stroke: colors[random(0, 5)],
      strokeWidth: random(1, 5)
    };
  }

  getTransitionData() {
    const n = random(4, 10)
    return range(n).map((i) => {
      return {
        x: i,
        y: random(2, 10)
      };
    });
  }

  componentDidMount() {
    setInterval(() => {
      this.setState({
        y: this.getYFunction(),
        style: this.getStyles(),
        transitionData: this.getTransitionData(),
        randomData: this.generateRandomData()
      });
    }, 3000);
  }
  render() {
    return (
      <ScrollView contentContainerStyle={styles.container}>
        <Text style={styles.text}>{"<VictoryPie/>"}</Text>
        <VictoryPie
          innerRadius={75}
          data={this.state.randomData}
          animate={{duration: 1500}}
        />

        <VictoryPie
          style={{
            labels: {
              fontSize: 15,
              padding: 75
            }
          }}
        />
        <VictoryPie
          style={{
            data: {
              stroke: "none",
              opacity: 0.3
            }
          }}
        />
        <VictoryPie innerRadius={90} />
        <VictoryPie
          endAngle={90}
          startAngle={-90}
        />
        <VictoryPie
          endAngle={90}
          innerRadius={90}
          padAngle={5}
          startAngle={-90}
        />
        <VictoryPie
          style={{
            labels: {
              fill: "white",
              stroke: "none",
              fontSize: 10,
              fontWeight: "bold"
            }
          }}
          data={[
            {x: "<5", y: 6279},
            {x: "5-13", y: 9182},
            {x: "14-17", y: 5511},
            {x: "18-24", y: 7164},
            {x: "25-44", y: 6716},
            {x: "45-64", y: 4263},
            {x: "≥65", y: 7502}
          ]}
          innerRadius={70}
          colorScale={[
            "#D85F49",
            "#F66D3B",
            "#D92E1D",
            "#D73C4C",
            "#FFAF59",
            "#E28300",
            "#F6A57F"
          ]}
        />
        <VictoryPie
          style={{
            data: {
              stroke: (data) => data.y > 75 ? "black" : "none",
              opacity: (data) => data.y > 75 ? 1 : 0.4
            }
          }}
          data={[
            {x: "Cat", y: 62},
            {x: "Dog", y: 91},
            {x: "Fish", y: 55},
            {x: "Bird", y: 55}
          ]}
        />

        <Text style={styles.text}>{"<VictoryChart/>"}</Text>

        <VictoryChart><VictoryBar/><VictoryLine/></VictoryChart>

        <VictoryChart><VictoryCandlestick data={candleData}/></VictoryChart>

        <VictoryChart domain={{x: [0, 4]}}>
          <VictoryGroup
            labels={["a", "b", "c"]}
            offset={10}
            colorScale={"qualitative"}
          >
            <VictoryBar
              data={[
                {x: 1, y: 1},
                {x: 2, y: 2},
                {x: 3, y: 5}
              ]}
            />
            <VictoryBar
              data={[
                {x: 1, y: 2},
                {x: 2, y: 1},
                {x: 3, y: 7}
              ]}
            />
            <VictoryBar
              data={[
                {x: 1, y: 3},
                {x: 2, y: 4},
                {x: 3, y: 9}
              ]}
            />
          </VictoryGroup>
        </VictoryChart>

        <VictoryChart>
          <VictoryScatter
            data={[
              {
                x: 1, y: 3, fill: "red",
                symbol: "plus", size: 6, label: "Red"
              },
              {
                x: 2, y: 5, fill: "magenta",
                size: 9, opacity: 0.4, label: "Magenta"
              },
              {
                x: 3, y: 4, fill: "orange",
                size: 5, label: "Orange"
              },
              {
                x: 4, y: 2, fill: "brown",
                symbol: "square", size: 6, label: "Brown"
              },
              {
                x: 5, y: 5, fill: "black",
                symbol: "triangleUp", size: 5, label: "Black"
              }
            ]}
          />
        </VictoryChart>

        <VictoryChart animate={{duration: 1500}}>
          <VictoryBar
            data={this.state.transitionData}
            style={{
              data: {
                fill: "tomato", width: 12
              }
            }}
            animate={{
              onExit: {
                duration: 500,
                before: () => ({
                  y: 0,
                  fill: "orange",
                  label: "BYE"
                })
              }
            }}
          />
        </VictoryChart>

        <VictoryChart>
          <VictoryStack>
            <VictoryArea
              data={[
                {x: "a", y: 2}, {x: "b", y: 3}, {x: "c", y: 5}, {x: "d", y: 4}, {x: "e", y: 7}
              ]}
            />
            <VictoryArea
              data={[
                {x: "a", y: 1}, {x: "b", y: 4}, {x: "c", y: 5}, {x: "d", y: 7}, {x: "e", y: 5}
              ]}
            />
            <VictoryArea
              data={[
                {x: "a", y: 3}, {x: "b", y: 2}, {x: "c", y: 6}, {x: "d", y: 2}, {x: "e", y: 6}
              ]}
            />
            <VictoryArea
              data={[
                {x: "a", y: 2}, {x: "b", y: 3}, {x: "c", y: 3}, {x: "d", y: 4}, {x: "e", y: 7}
              ]}
            />
          </VictoryStack>
        </VictoryChart>

        <Text style={styles.text}>{"<VictoryLine />"}</Text>

        <VictoryLine />

        <VictoryLine
          data={[
            {x: 0, y: 1},
            {x: 1, y: 3},
            {x: 2, y: 2},
            {x: 3, y: 4},
            {x: 4, y: 3},
            {x: 5, y: 5}
          ]}
        />

        <VictoryLine
          data={[
            {amount: 1, yield: 1, error: 0.5},
            {amount: 2, yield: 2, error: 1.1},
            {amount: 3, yield: 3, error: 0},
            {amount: 4, yield: 2, error: 0.1},
            {amount: 5, yield: 1, error: 1.5}
          ]}
          x={"amount"}
          y={(data) => (data.yield + data.error)}
        />

        <VictoryLine y={(data) => Math.sin(2 * Math.PI * data.x)} />

        <VictoryLine
          height={300}
          domain={[0, 5]}
          padding={75}
          data={[
            {x: 0, y: 1},
            {x: 1, y: 3},
            {x: 2, y: 2},
            {x: 3, y: 4},
            {x: 4, y: 3},
            {x: 5, y: 5}
          ]}
          interpolation="cardinal"
          label="LINE"
          style={{
            data: {
              stroke: "#822722",
              strokeWidth: 3
            },
            labels: {fontSize: 12}
          }}
        />

        <VictoryLine
          width={300}
          style={{
            data: {
              stroke: (data) => {
                const y = data.map((d) => d.y);
                return Math.max(...y) > 2 ?
                  "red" : "blue";
              }
            }
          }}
          data={[
            {x: 0, y: 1},
            {x: 1, y: 3},
            {x: 2, y: 2},
            {x: 3, y: 4},
            {x: 4, y: 3},
            {x: 5, y: 5}
          ]}
        />

        <VictoryLine
          style={{
            data: {stroke: "red", strokeWidth: 9}
          }}
          interpolation={"linear"}
          data={[
            {x: 0, y: 1},
            {x: 1, y: 3},
            {x: 2, y: 2},
            {x: 3, y: 4},
            {x: 4, y: 3},
            {x: 5, y: 5}
          ]}
        />

        <VictoryLine
          style={{data: this.state.style}}
          interpolation="basis"
          animate={{duration: 1500}}
          y={this.state.y}
        />

        <Text style={styles.text}>{"<VictoryArea />"}</Text>

        <VictoryArea />

        <VictoryArea
          data={[
            {x: 1, y: 1},
            {x: 2, y: 2},
            {x: 3, y: 3},
            {x: 4, y: 1},
            {x: 5, y: 3},
            {x: 6, y: 4},
            {x: 7, y: 2}
          ]}
        />

        <VictoryArea
          data={[
            {amount: 1, yield: 1, error: 0.5},
            {amount: 2, yield: 2, error: 1.1},
            {amount: 3, yield: 3, error: 0},
            {amount: 4, yield: 2, error: 0.1},
            {amount: 5, yield: 1, error: 1.5}
          ]}
          x={"amount"}
          y={(data) => (data.yield + data.error)}
        />

        <VictoryGroup
          width={300}
          height={375}
          style={{data: {opacity: 0.3}}}
        >
          <VictoryArea
            data={[
              {x: 1, y: 1},
              {x: 2, y: 2},
              {x: 3, y: 3}
            ]}
          />
          <VictoryArea
            data={[
              {x: 1, y: 2},
              {x: 2, y: 1},
              {x: 3, y: 1}
            ]}
          />
          <VictoryArea
            data={[
              {x: 1, y: 3},
              {x: 2, y: 4},
              {x: 3, y: 2}
            ]}
          />
        </VictoryGroup>

        <VictoryStack
          width={300}
          height={375}
        >
          <VictoryArea
            data={[
              {x: 1, y: 1},
              {x: 2, y: 2},
              {x: 3, y: 3}
            ]}
          />
          <VictoryArea
            data={[
              {x: 1, y: 2},
              {x: 2, y: 1},
              {x: 3, y: 1}
            ]}
          />
          <VictoryArea
            data={[
              {x: 1, y: 3},
              {x: 2, y: 4},
              {x: 3, y: 2}
            ]}
          />
        </VictoryStack>

        <VictoryStack
          width={300}
          height={450}
          style={{ data: {
            strokeDasharray: "5,5",
            strokeWidth: 2,
            fillOpacity: 0.4
          }}}
        >
          <VictoryArea
            style={{ data: {
              fill: "tomato", stroke: "tomato"
            }}}
            data={[
              {x: 1, y: 1},
              {x: 2, y: 2},
              {x: 3, y: 3}
            ]}
          />
          <VictoryArea
            style={{ data: {
              fill: "orange", stroke: "orange"
            }}}
            data={[
              {x: 1, y: 2},
              {x: 2, y: 1},
              {x: 3, y: 1}
            ]}
          />
          <VictoryArea
            style={{ data: {
              fill: "gold", stroke: "gold"
            }}}
            data={[
              {x: 1, y: 3},
              {x: 2, y: 4},
              {x: 3, y: 2}
            ]}
          />
        </VictoryStack>

        <Text style={styles.text}>{"<VictoryBar />"}</Text>

        <VictoryBar />

        <VictoryBar
          data={[
            {x: 1, y: 1},
            {x: 2, y: 2},
            {x: 3, y: 3},
            {x: 4, y: 2},
            {x: 5, y: 1}
          ]}
        />

        <VictoryGroup
          width={300}
          height={375}
          offset={20}
          colorScale={"qualitative"}
        >
          <VictoryBar
            data={[
              {x: 1, y: 1},
              {x: 2, y: 2},
              {x: 3, y: 3}
            ]}
          />
          <VictoryBar
            data={[
              {x: 1, y: 2},
              {x: 2, y: 1},
              {x: 3, y: 1}
            ]}
          />
          <VictoryBar
            data={[
              {x: 1, y: 3},
              {x: 2, y: 4},
              {x: 3, y: 2}
            ]}
          />
        </VictoryGroup>

        <VictoryStack
          width={300}
          height={375}
          colorScale={"qualitative"}
        >
          <VictoryBar
            data={[
              {x: 1, y: 1},
              {x: 2, y: 2},
              {x: 3, y: 3}
            ]}
          />
          <VictoryBar
            data={[
              {x: 1, y: 2},
              {x: 2, y: 1},
              {x: 3, y: 1}
            ]}
          />
          <VictoryBar
            data={[
              {x: 1, y: 3},
              {x: 2, y: 4},
              {x: 3, y: 2}
            ]}
          />
        </VictoryStack>

        <VictoryBar
          height={375}
          padding={75}
          style={{
            data: {
              fill: (data) => data.y > 2 ? "red" : "blue"
            }
          }}
          data={[
            {x: 1, y: 1},
            {x: 2, y: 2},
            {x: 3, y: 3},
            {x: 4, y: 2},
            {x: 5, y: 1}
          ]}
        />

        <Text style={styles.text}>{"<VictoryScatter />"}</Text>

        <VictoryScatter />

        <VictoryScatter
          y={(data) =>
            Math.sin(2 * Math.PI * data.x)
          }
        />

        <VictoryScatter
          data={[
            {x: 1, y: 3},
            {x: 2, y: 5},
            {x: 3, y: 4},
            {x: 4, y: 2},
            {x: 5, y: 5}
          ]}
          size={8}
          symbol="star"
          style={{
            data: {
              fill: "gold",
              stroke: "orange",
              strokeWidth: 3
            }
          }}
        />

        <VictoryScatter
          style={{
            data: {
              fill: (data) => data.y > 0 ?
                "red" : "blue"
            }
          }}
          symbol={(data) => data.y > 0 ?
            "triangleUp" : "triangleDown"
          }
          y={(d) => Math.sin(2 * Math.PI * d.x)}
          samples={25}
        />

        <Text style={styles.text}>{"<VictoryAxis />"}</Text>

        <VictoryAxis height={100} />

        <VictoryAxis
          height={100}
          scale="time"
          tickValues={[
            new Date(1980, 1, 1),
            new Date(1990, 1, 1),
            new Date(2000, 1, 1),
            new Date(2010, 1, 1),
            new Date(2020, 1, 1)
          ]}
          tickFormat={(x) => x.getFullYear()}
        />

        <Svg width={320} height={320}>
          <VictoryAxis
            width={320}
            height={320}
            domain={[-10, 10]}
            crossAxis={true}
            offsetY={160}
            standalone={false}/>
          <VictoryAxis dependentAxis
            width={320}
            height={320}
            domain={[-10, 10]}
            crossAxis={true}
            offsetX={160}
            standalone={false}/>
        </Svg>

        <VictoryAxis
          dependentAxis
          padding={{left: 50, top: 20, bottom: 20}}
          scale="log"
          domain={[1, 5]}
        />

        <Text style={styles.text}>{"<VictoryErrorBar />"}</Text>

        <VictoryErrorBar
          style={{
            data: {stroke: "red", strokeWidth: 2}
          }}
          data={[
            {x: 1, y: 1, errorX: [1, 0.5], errorY: .1},
            {x: 2, y: 2, errorX: [1, 3], errorY: .1},
            {x: 3, y: 3, errorX: [1, 3], errorY: [.2, .3]},
            {x: 4, y: 2, errorX: [1, 0.5], errorY: .1},
            {x: 5, y: 1, errorX: [1, 0.5], errorY: .2}
          ]}
        />

        <Text style={styles.text}>{"Tooltips"}</Text>
        <VictoryChart
          domain={{y: [-25, 25]}}
        >
          <VictoryGroup
            data={
              range(10).map((i) => {
                return {
                  x: i,
                  y: random(-20, 20)
                };
              })
            }
          >
            <VictoryLine/>
            <VictoryVoronoiTooltip
              labels={(d) => `x: ${d.x} \n y: ${d.y}`}
            />
          </VictoryGroup>
        </VictoryChart>

        <VictoryChart>
          <VictoryScatter
            labelComponent={<VictoryTooltip/>}
            data={[
              {
                x: 1, y: 3, fill: "red",
                symbol: "plus", size: 6, label: "WOW\nSTUFF"
              },
              {
                x: 2, y: 5, fill: "magenta",
                size: 9, opacity: 0.4, label: "WAT"
              },
              {
                x: 3, y: 4, fill: "orange",
                size: 5, label: "LABEL"
              },
              {
                x: 4, y: 2, fill: "brown",
                symbol: "square", size: 6, label: "OKAY"
              },
              {
                x: 5, y: 5, fill: "black",
                symbol: "triangleUp", size: 5, label: "GOOD"
              }
            ]}
          />
        </VictoryChart>
      </ScrollView>
    );
  }
}

AppRegistry.registerComponent('VictoryDemo', () => Demo);
